protocols/autonatv2/src/client/behaviour.rs

```diff
@@ -0,0 +1,228 @@
+                    ..
+                },
+            suspicious_addrs,
+            successfull_addr,
+        }: ResponseInfo,
+    ) -> Result<ReachabilityTestSucc, ReachabilityTestErr> {
+        match (status, dial_status) {
+            (ResponseStatus::E_REQUEST_REJECTED, _) => {
+                Err(ReachabilityTestErr::ServerRejectedDialRequest { peer_id })
+            }
+            (ResponseStatus::E_DIAL_REFUSED, _) => {
+                Err(ReachabilityTestErr::ServerChoseNotToDialAnyAddress { peer_id })
+            }
+            (ResponseStatus::E_INTERNAL_ERROR, _) => {
+                Err(ReachabilityTestErr::InternalServerError { peer_id })
+            }
+            (ResponseStatus::OK, DialStatus::UNUSED) => {
+                Err(ReachabilityTestErr::InvalidResponse { peer_id })
+            }
+            (ResponseStatus::OK, DialStatus::E_DIAL_ERROR) => {
+                Err(ReachabilityTestErr::UnableToConnectOnSelectedAddress {
+                    peer_id,
+                    addr: successfull_addr,
+                })
+            }
+            (ResponseStatus::OK, DialStatus::E_DIAL_BACK_ERROR) => {
+                Err(ReachabilityTestErr::FailureDuringDialBack {
+                    peer_id,
+                    addr: successfull_addr,
+                })
+            }
+            (ResponseStatus::OK, DialStatus::OK) => Ok(ReachabilityTestSucc {
+                server_peer: peer_id,
+                visible_addr: successfull_addr,
+                suspicious_addrs,
+            }),
+        }
+    }
+
+    fn handle_command(&mut self, command: Command) {
+        match command {
+            Command::TestListenerReachability {
+                server_peer,
+                local_addrs,
+            } => {
+                let cleaned_local_addrs = local_addrs.iter().filter(|addr| {
+                    !addr.iter().any(|p| match p {
+                        Protocol::Ip4(ip) if !IpExt::is_global(&ip) => true,
+                        Protocol::Ip6(ip) if !IpExt::is_global(&ip) => true,
+                        Protocol::Dns(m)
+                        | Protocol::Dns4(m)
+                        | Protocol::Dns6(m)
+                        | Protocol::Dnsaddr(m) => m == "localhost" || m.ends_with(".local"),
+                        _ => false,
+                    })
+                });
+                let dial_opts = DialOpts::peer_id(server_peer.clone()).build();
+                let dial_event = ToSwarm::Dial { opts: dial_opts };
+                self.pending_events.push_back(dial_event);
+                let dial_request = DialRequest {
+                    addrs: local_addrs,
+                    nonce: self.rng.next_u64(),
+                };
+                self.pending_events.push_back(ToSwarm::NotifyHandler {
+                    peer_id: server_peer,
+                    handler: NotifyHandler::Any,
+                    event: FromBehaviour::Dial(dial_request),
+                })
```

Comment:
This will be silently dropped because there is no handler yet and the `Dial` you added before is not instant. I'd instead recommend to not emit `Dial` at all and instead send the dial request on a connection where we know that the remote supports the autonat protocol. You can learn this from `ConnectionEvent::RemoteProtocolsChanged`.

We can also include a dedicated function on `Behaviour` that allows a user to "force-add" an autonat server that I'd make that an additional feature on top. That would be useful for users that don't use the identify or similar protocol.

Reply:
I think this is addressed, correct me if I'm wrong.